[Don't Merge] Pulsar 2.11.0 release notes (#167)
apache/pulsar-site: site2/website-next/release-notes/versioned/pulsar-2.11.0.md

- Fix incorrect entryId in warn log when reading entry from tiered storage [#14685](https://github.com/apache/pulsar/pull/14685)
- Optimize PositionImpl toString, compareTo and hashCode methods [#14678](https://github.com/apache/pulsar/pull/14678)
- Improve readability of ManagedCursorImpl source code [#14617](https://github.com/apache/pulsar/pull/14617)
- Intercept ACK command by BrokerInterceptor [#14616](https://github.com/apache/pulsar/pull/14616)
- Add filteredMsgCount for pulsar-admin broker-stats topics [#14531](https://github.com/apache/pulsar/pull/14531)
- Fix NPE when subscription is already removed [#14363](https://github.com/apache/pulsar/pull/14363)
- Fix backpressure handling in Jetty web server configuration [#14353](https://github.com/apache/pulsar/pull/14353)
- Optimize find nics process [#14340](https://github.com/apache/pulsar/pull/14340)
- Support pass http auth status [#14044](https://github.com/apache/pulsar/pull/14044)
- Support caching to drain backlog consumers [#12258](https://github.com/apache/pulsar/pull/12258)
- Strict bookie affinity group strategy [#12025](https://github.com/apache/pulsar/pull/12025)

> ### Clients

Reviewer: Is this part (line 217-285) already split into the following independent release notes? If so, should this part be removed here?

<img width="206" alt="image" src="https://user-images.githubusercontent.com/50226895/185337812-c1721ed6-dfd5-4915-971a-3aa886e31ac0.png">

- [Pulsar CLI] Support getEarliestTimeInBacklog at getPartitionedStats method [#16388](https://github.com/apache/pulsar/pull/16388)
- [Pulsar CLI] Add query options to the get topics in Namespace [#16167](https://github.com/apache/pulsar/pull/16167)
- [Pulsar CLI] Add cli cmd for subscription level dispatch-rate-limiter [#15862](https://github.com/apache/pulsar/pull/15862)
- [Pulsar CLI] Disable Pulsar client memory limit by default [#15748](https://github.com/apache/pulsar/pull/15748)
- [Pulsar CLI] Fix check on create function class name [#15700](https://github.com/apache/pulsar/pull/15862)
- [Pulsar CLI] Support filtering system topic when get list [#15410](https://github.com/apache/pulsar/pull/15410)
- [Pulsar CLI] Change the default for max-connections from 100 to 1 [#15387](https://github.com/apache/pulsar/pull/15387)
- [Pulsar CLI] Add short name for full name in admin cli [#14301](https://github.com/apache/pulsar/pull/14301)
- [Client Tool] Add tlsTrustCertFilePath as CLI argument for pulsar-client-tool [#16961](https://github.com/apache/pulsar/pull/16961)
- [Perf Tool] Able to set maxLookupRequest for pulsar-perf [#16967](https://github.com/apache/pulsar/pull/16967)

> ### Others

Reviewer: Should this part be split into several parts to improve readbility?

eg.
```
#### Proxy

#### Schema

#### Tiered storage

...
```

- Bump os-maven-plugin version from 1.4.1.Final to 1.7.0 [#16308](https://github.com/apache/pulsar/pull/16308)
- Bump presto.version to run PrestoServer on JDK11+ [#16163](https://github.com/apache/pulsar/pull/16163)
- Update fastjson version to 1.2.83 [#16148](https://github.com/apache/pulsar/pull/16148)
- Bump pyyaml from 5.3.1 to 5.4.1 to solve CVE-2020-14343 [#15989](https://github.com/apache/pulsar/pull/15989)
- Bump spring version to 5.3.20 to solve CVE-2022-22970 [#15699](https://github.com/apache/pulsar/pull/15699)
- Bump snappy zstd version to fix CompressorCodecBackwardCompatTest on Apple M1 [#15698](https://github.com/apache/pulsar/pull/15698)
- Athenz: do not use uber-jar and bump to 1.10.50 to remove jackson-databind shaded dependency [#14884](https://github.com/apache/pulsar/pull/14884)
- Remove --illegal-access errors resulting from Google Guice - Pulsar IO, Offloaders and Pulsar SQL - Bump Guice to 5.1.0 [#14300](https://github.com/apache/pulsar/pull/14300)
- Bump prometheus client version from 0.5.0 to 0.15.0 [#13785](https://github.com/apache/pulsar/pull/13785)
- Upgrade log4j2 version to 2.18.0 [#16884](https://github.com/apache/pulsar/pull/16884

> ### Documentation

Reviewer: Items in some chapters use this format: `[xxx][xxx] xxxx` 
While items in the `Documentation` chapter do not use this format. 
Should we keep consistent？

Reviewer 2: These two fields can improve the readability of a single PR in GitHub, but go the opposite in the release notes IMHO, especially the second one because all the items under a section usually share the same scope.

---
id: pulsar-2.11.0
title: Apache Pulsar 2.11.0 
sidebar_label: Apache Pulsar 2.11.0 
---

#### 2022-08-16

### Important notice
- Rename Pulsar txn metrics to specify OpenMetrics [#16581](https://github.com/apache/pulsar/pull/16581)
- Rename Pulsar schema metrics to specify OpenMetrics [#16610](https://github.com/apache/pulsar/pull/16610)
- Rename Pulsar lb metrics to specify OpenMetrics [#16611](https://github.com/apache/pulsar/pull/16611)
- Remove jul-to-slf4j #16320 [#16320](https://github.com/apache/pulsar/pull/16320)
- Fix get non-persistent topics issue in Namespaces [#16170](https://github.com/apache/pulsar/pull/16170)
- Fix producer/consume permission can’t get schema [#15956](https://github.com/apache/pulsar/pull/15956)
- Add verification for configured default backlog quota and retention [#15441](https://github.com/apache/pulsar/pull/15441)

> - Fix broker LoadBalance uneffective [#15314](https://github.com/apache/pulsar/pull/15314)

Reviewer 2: ```suggestion
- Fix ineffective broker LoadBalance [#15314](https://github.com/apache/pulsar/pull/15314)
```

- Ensure the deletion consistency of topic and schema [#14608](https://github.com/apache/pulsar/pull/14608)
- Transaction cumulative ack redeliver change [#14371](https://github.com/apache/pulsar/pull/14371)
- Bump prometheus client version from 0.5.0 to 0.15.0 [#13785](https://github.com/apache/pulsar/pull/13785)
- [PIP-158] Split client TLS transport encryption from authentication [#15634](https://github.com/apache/pulsar/pull/15634)

> - [PIP-155] Removed Python 2 support [#15376](https://github.com/apache/pulsar/pull/15376)

Reviewer 2: ```suggestion
- [PIP-155] Remove Python 2 support [#15376](https://github.com/apache/pulsar/pull/15376)
```

- [PIP-156] Build Pulsar Server on Java 17 [#15264](https://github.com/apache/pulsar/pull/15264)
- [PIP-162] Enable system topic by default [#15619](https://github.com/apache/pulsar/pull/15619)
- [PIP-174] New managed ledger entry cache implementation [#15955](https://github.com/apache/pulsar/pull/15955)

### PIPs
- [PIP 37] Support chunking with Shared subscription [#16202](https://github.com/apache/pulsar/pull/16202)
- [PIP-74] Pulsar client memory limits [#14400](https://github.com/apache/pulsar/pull/14400)
- [PIP-81] Split createNewMetadataLedger into multiple methods for reuse [#15425](https://github.com/apache/pulsar/pull/15425)
- [PIP 97] Update Authentication Interfaces to Include Async Authentication Methods [#12104](https://github.com/apache/pulsar/pull/12104)
- [PIP-100] Support pluggable topic factory [#12235](https://github.com/apache/pulsar/pull/12235)
- [PIP-105] Store Subscription properties [#15757](https://github.com/apache/pulsar/pull/15757)
- [PIP-117] Change Pulsar standalone defaults [#15478](https://github.com/apache/pulsar/pull/15478)

> - [PIP-132] Include message header size when check maxMessageSize for non-batch message on the client side. [#14007](https://github.com/apache/pulsar/pull/14007)

Reviewer 2: ```suggestion
- [PIP-132] Include message header size when checking maxMessageSize for non-batch messages on the client side. [#14007](https://github.com/apache/pulsar/pull/14007)
```

- [PIP-136] Sync Pulsar metadata across multiple clouds [#16425](https://github.com/apache/pulsar/pull/16425)
- [PIP-143] Support split bundle by specified boundaries [#13796](https://github.com/apache/pulsar/pull/13796)
- [PIP-145] Improve performance of regex subscriptions [#14505](https://github.com/apache/pulsar/pull/14505)
- [PIP-146] ManagedCursorInfo compression [#14542](https://github.com/apache/pulsar/pull/14542)

> - [PIP 149] Making the REST Admin API fully async [#14365](https://github.com/apache/pulsar/issues/14365)

Reviewer 2: ```suggestion
- [PIP 149] Make the REST Admin API fully async [#14365](https://github.com/apache/pulsar/issues/14365)
```

- [PIP-152] Support subscription level dispatch rate limiter setting [#15295](https://github.com/apache/pulsar/issues/15295)
- [PIP-153] Optimize metadataPositions in MLPendingAckStore [#15137](https://github.com/apache/pulsar/pull/15137)
- [PIP-160] Make transactions work more efficiently by aggregation operation for transaction log and pending ack store [#15370](https://github.com/apache/pulsar/issues/15370)
- [PIP-161] Exclusive Producer: ability to fence out an existing Producer [#15488](https://github.com/apache/pulsar/pull/15488)
- [PIP-163] Add lowWaterMark check before appending entry to TB [#15424](https://github.com/apache/pulsar/pull/15424)

> - [PIP-166] Function add MANUAL delivery semantics [#16279](https://github.com/apache/pulsar/pull/16279)

Reviewer 2: ```suggestion
- [PIP-166] Add MANUAL delivery semantics for Functions [#16279](https://github.com/apache/pulsar/pull/16279)
```

- [PIP-179] Support the admin API to check unknown request parameters [#16135](https://github.com/apache/pulsar/issues/16135)
- [PIP-181] Pulsar Shell [#16250](https://github.com/apache/pulsar/issues/16250)
- [PIP-184] Topic specific consumer priorityLevel [#16715](https://github.com/apache/pulsar/issues/16715)

> - [PIP-187] Add API to analyze a subscription backlog and provide a accurate value [#16545](https://github.com/apache/pulsar/issues/16545)

Reviewer 2: ```suggestion
- [PIP-187] Add API to analyze a subscription backlog and provide an accurate value [#16545](https://github.com/apache/pulsar/issues/16545)
```

  - Enable evaluating subscription pattern on broker side [#14804](https://github.com/apache/pulsar/pull/14804)
  - Notifications for faster topic discovery [#16062](https://github.com/apache/pulsar/pull/16062)
- [PIP-156] Build Pulsar Server on Java 17
  - Build changes by dropping some required checks [#15496](https://github.com/apache/pulsar/pull/15496)
  - Build Pulsar Server on Java 17 [#15264](https://github.com/apache/pulsar/pull/15264)
- Optimize getting ledger and entry id from entry [#17108](https://github.com/apache/pulsar/pull/17108)
- Pass subscriptionName to auth service [#17123](https://github.com/apache/pulsar/pull/17123)
- Support loadBalancerSheddingIntervalMinutes dynamic configuration [#16408](https://github.com/apache/pulsar/pull/16408)
- Fix out of order data replication [#17154](https://github.com/apache/pulsar/pull/17154)
- Fix schema does not replicate successfully [#17049](https://github.com/apache/pulsar/pull/17049)
- Streaming dispatcher stuck after reading the first entry with SHARED subscriptions [#17143](https://github.com/apache/pulsar/pull/17143)
- Fix calculate avg message per entry [#17046](https://github.com/apache/pulsar/pull/17046)

> - Bundle-data metadata leak because of bundlestats was not clean [#17095](https://github.com/apache/pulsar/pull/17095)

Reviewer 2: ```suggestion
- Fix bundle-data metadata leak because bundle stats was not cleaned up [#17095](https://github.com/apache/pulsar/pull/17095)
```

- Duplicate ByteBuffer when Caching Backlogged Consumers [#17105](https://github.com/apache/pulsar/pull/17105)
- Fix offload read handle NPE [#17056](https://github.com/apache/pulsar/pull/17056)
- Move the state check forward [#17020](https://github.com/apache/pulsar/pull/17020)
- Improve naming for delete topic error [#16965](https://github.com/apache/pulsar/pull/16965)
- Increment topic stats outbound message counters and update rate after messages have been written to the TCP/IP connection [#17043](https://github.com/apache/pulsar/pull/17043)
- Prevent StackOverFlowException in SHARED subscription [#16968](https://github.com/apache/pulsar/pull/16968)
- Improve the package download process [#16365](https://github.com/apache/pulsar/pull/16365)
- Skip mis-configured resource usage(>100%) in load balancer [#16937](https://github.com/apache/pulsar/pull/16937)

> - Not allow to terminate system topic [#17006](https://github.com/apache/pulsar/pull/17006)

Reviewer 2: ```suggestion
- Not allow terminating system topic [#17006](https://github.com/apache/pulsar/pull/17006)
```

- Fixed error when delayed messages trackers state grows to >1.5GB [#16490](https://github.com/apache/pulsar/pull/16490)
- Fix RawReader hasMessageAvailable returns true when no messages [#16443](https://github.com/apache/pulsar/pull/16443)
- Fix No such ledger exception [#16420](https://github.com/apache/pulsar/pull/16420)
- Add config maxUnloadBundleNumPerShedding for UniformLoadShedder [#16409](https://github.com/apache/pulsar/pull/16409)
- Fix etcd cluster error and add test for etcd cluster [#16309](https://github.com/apache/pulsar/pull/16309)
- Skip reading more entries for a pending read with no more entries [#16400](https://github.com/apache/pulsar/pull/16400)
- Recycle OpReadEntry in some corner cases [#16399](https://github.com/apache/pulsar/pull/16399)
- Add dynamic configuration for UniformLoadShedder [#16391](https://github.com/apache/pulsar/pull/16391)
- Fix RawReader out of order [#16390](https://github.com/apache/pulsar/pull/16390)
- Create the cursor ledger lazily to improve the subscribe performance [#16389](https://github.com/apache/pulsar/pull/16389)
- Release the entry in getEarliestMessagePublishTimeOfPos [#16386](https://github.com/apache/pulsar/pull/16386)

> - Update cpu ResourceUsage before updating SystemResourceUsage usage [#16366](https://github.com/apache/pulsar/pull/16366)

Reviewer 2: ```suggestion
- Update CPU ResourceUsage before updating SystemResourceUsage usage [#16366](https://github.com/apache/pulsar/pull/16366)
```

- Fix the broker close hanged issue [#15755](https://github.com/apache/pulsar/pull/15755)
- Disable memory limit controller in internal Pulsar clients [#15752](https://github.com/apache/pulsar/pull/15752)
- Add timeout for unload namespace bundle [#15719](https://github.com/apache/pulsar/pull/15719)
- Support schemaValidationEnforced on topic level [#15712](https://github.com/apache/pulsar/pull/15712)
- Fix NPE when put value to RangeCache [#15707](https://github.com/apache/pulsar/pull/15707)
- Fast return if ack cumulative illegal [#15695](https://github.com/apache/pulsar/pull/15695)
- Fix creating producer failure when set backlog quota [#15663](https://github.com/apache/pulsar/pull/15663)
- Fix logging in shutdown when broker shutdown [#15626](https://github.com/apache/pulsar/pull/15626)
- Fix MetadataStoreException$NotFoundException while doing topic lookup [#15633](https://github.com/apache/pulsar/pull/15633)
- Use dispatchRateLimiterLock to update dispatchRateLimiter [#15601](https://github.com/apache/pulsar/pull/15601)
- Add .yaml suffix support for broker filter config file [#15600](https://github.com/apache/pulsar/pull/15600)

> - Added newline to end of standalone.conf [#15591](https://github.com/apache/pulsar/pull/15591)

Reviewer 2: ```suggestion
- Add newline to the end of standalone.conf [#15591](https://github.com/apache/pulsar/pull/15591)
```

- Close publishLimiter when disable it [#15520](https://github.com/apache/pulsar/pull/15520)
- Enable SO_REUSEADDR on listen sockets for broker,proxy [#15517](https://github.com/apache/pulsar/pull/15517)
- Fix to avoid TopicStatsImpl NPE even if producerName is null [#15502](https://github.com/apache/pulsar/pull/15502)
- Close connection if a ping or pong message cannot be sent [#15382](https://github.com/apache/pulsar/pull/15382)
- Introduce BitRateUnit for calculate bitrate [#15435](https://github.com/apache/pulsar/pull/15435)
- Add verification for configured default backlog quota and retention [#15441](https://github.com/apache/pulsar/pull/15441)
- Change unnecessary atomic to basic type [#15369](https://github.com/apache/pulsar/pull/15369)
- Support shrink for ConcurrentSortedLongPairSet [#15354](https://github.com/apache/pulsar/pull/15354)
- Use shrink map for message redelivery [#15342](https://github.com/apache/pulsar/pull/15342)
- Fix precision error in UsageUnit [#15322](https://github.com/apache/pulsar/pull/15322)
- Optimized namespace-level dispatcherRateLimiter update [#15315](https://github.com/apache/pulsar/pull/15315)

> - Fix broker LoadBalance uneffective [#15314](https://github.com/apache/pulsar/pull/15314)

Reviewer 2: ```suggestion
- Fix ineffective broker LoadBalance [#15314](https://github.com/apache/pulsar/pull/15314)
```

> - Fix wrong unit of NIC speed on linux [#15304](https://github.com/apache/pulsar/pull/15304)

Reviewer 2: ```suggestion
- Fix wrong unit of NIC speed on Linux [#15304](https://github.com/apache/pulsar/pull/15304)
```

- Skip unnecessary DNS resolution when creating AuthenticationDataHttp instance [#15221](https://github.com/apache/pulsar/pull/15221)
- Fix MessageDeduplication#inactiveProducers may not be persistence correctly [#15206](https://github.com/apache/pulsar/pull/15206)
- Cancel fencedTopicMonitoringTask when topic closed normally [#15202](https://github.com/apache/pulsar/pull/15202)
- Fix npe when doCacheEviction [#15184](https://github.com/apache/pulsar/pull/15184)
- Make health check fail if dead locked threads are detected [#15155](https://github.com/apache/pulsar/pull/15155)
- Fix parameter saslJaasBrokerSectionName in broker.conf [#15110](https://github.com/apache/pulsar/pull/15110)
- Fix potential to add duplicated consumer [#15051](https://github.com/apache/pulsar/pull/15051)
- Fix rewind failed when redeliverUnacknowledgedMessages [#15046](https://github.com/apache/pulsar/pull/15046)
- Fix race condition in updating lastMarkDeleteEntry field [#15031](https://github.com/apache/pulsar/pull/15031)
- Avoid heartbeat topic to offload [#15008](https://github.com/apache/pulsar/pull/15008)

> - Evicting cache data by the slowest markDeletedPosition [#14985](https://github.com/apache/pulsar/pull/14985)

Reviewer 2: ```suggestion
- Evict cache data according to the slowest markDeletedPosition [#14985](https://github.com/apache/pulsar/pull/14985)
```

- Add filteredMsgCount for pulsar-admin broker-stats topics [#14531](https://github.com/apache/pulsar/pull/14531)
- Fix NPE when subscription is already removed [#14363](https://github.com/apache/pulsar/pull/14363)
- Fix backpressure handling in Jetty web server configuration [#14353](https://github.com/apache/pulsar/pull/14353)
- Optimize find nics process [#14340](https://github.com/apache/pulsar/pull/14340)
- Support pass http auth status [#14044](https://github.com/apache/pulsar/pull/14044)
- Support caching to drain backlog consumers [#12258](https://github.com/apache/pulsar/pull/12258)
- Strict bookie affinity group strategy [#12025](https://github.com/apache/pulsar/pull/12025)

### Clients
- [PIP 74] Pulsar client memory limits
  - Support auto scaled consumer receiver queue [#14494](https://github.com/apache/pulsar/pull/14494)
  - Support dynamic limit of consumer receiver queue [#14400](https://github.com/apache/pulsar/pull/14400)
  - Support consumer client memory limit [#15216](https://github.com/apache/pulsar/pull/15216)
- [Java] LastBatchSendNanoTime initialization [#17058](https://github.com/apache/pulsar/pull/17058)

> - [Java] Fixed cnx channel Inactive causing the request fail to time out and fail to return [#17051](https://github.com/apache/pulsar/pull/17051)

Reviewer 2: ```suggestion
- [Java] Fix inactive cnx channel causing the request to fail to time out and fail to return [#17051](https://github.com/apache/pulsar/pull/17051)
```

- [Java] Release semaphore before discarding messages in batchMessageContainer [#17019](https://github.com/apache/pulsar/pull/17019)
- [Java] Reduce code duplication in admin client [#16377](https://github.com/apache/pulsar/pull/16377)
- [Java] Remove redundant check for chunked message TotalChunkMsgSize in ConsumerImpl [#16797](https://github.com/apache/pulsar/pull/16797)
- [Java] Fix client memory limit currentUsage leak and semaphore release duplicated in ProducerImpl [#16837](https://github.com/apache/pulsar/pull/16837)

> - [Java] Forget to update memory usage when invalid message [#16835](https://github.com/apache/pulsar/pull/16835)

Reviewer 2: ```suggestion
- [Java] Release memory usage for invalid messages [#16835](https://github.com/apache/pulsar/pull/16835)
```

- [Java] Fix subscription topic name error [#16719](https://github.com/apache/pulsar/pull/16719)
- [Java] Send CloseConsumer on client timeout [#16616](https://github.com/apache/pulsar/pull/16616)
- [Java] Add message key if exists to deadLetter messages [#16615](https://github.com/apache/pulsar/pull/16615)
- [Java] Make DeadLetterPolicy deserializable [#16513](https://github.com/apache/pulsar/pull/16513)
- [Java] Improve performance of multi-topic consumer with more than one IO thread [#16336](https://github.com/apache/pulsar/pull/16336)
- [Java] Add initialization for the OpSendMsg [#16256](https://github.com/apache/pulsar/pull/16256)
- [Java] Replace ScheduledExecutor to improve performance of message consumption [#16236](https://github.com/apache/pulsar/pull/16236)

> - [Java] Fix large message sometimes cannot be split into chunks [#16196](https://github.com/apache/pulsar/pull/16196)

Reviewer 2: ```suggestion
- [Java] Fix large messages sometimes cannot be split into chunks [#16196](https://github.com/apache/pulsar/pull/16196)
```

- [Java] Only trigger the batch receive timeout when having pending batch receives requests [#16160](https://github.com/apache/pulsar/pull/16160)
- [Java] Fix the startMessageId can't be respected as the ChunkMessageID [#16154](https://github.com/apache/pulsar/pull/16154)
- [Java] Fix auto cluster failover can't resolve host bug [#16152](https://github.com/apache/pulsar/pull/16152)
- [Java] Switch to rely on Netty for Hostname Verification [#15824](https://github.com/apache/pulsar/pull/15824)

> - [Java] Prevent the trigger running concurrently in the memory limit controller [#15737](https://github.com/apache/pulsar/pull/15737)

Reviewer 2: ```suggestion
- [Java] Prevent the trigger from running concurrently in the memory limit controller [#15737](https://github.com/apache/pulsar/pull/15737)
```

- [Java] Fix messages sent by producers without schema cannot be decoded [#15622](https://github.com/apache/pulsar/pull/15622)
- [Java] Remove sensitive msg from consumer/producer stats log [#15483](https://github.com/apache/pulsar/pull/15483)

> - [Java] Add pending messages information while print the producer stats [#15440](https://github.com/apache/pulsar/pull/15440)

Reviewer 2: ```suggestion
- [Java] Add pending messages information while printing the producer stats [#15440](https://github.com/apache/pulsar/pull/15440)
```

- [Java] Fix flaky BatchMessageTest by initializing lastBatchSendNanoTime [#15406](https://github.com/apache/pulsar/pull/15406)
- [Java] Check consumer schema null in advance [#15327](https://github.com/apache/pulsar/pull/15327)
- [Java] Fix inconsistent parameter of TopicPolicies.getSubscriptionDispatchRate [#15293](https://github.com/apache/pulsar/pull/15293)
- [Java] Fix typo in ConsumerBuilder [#15194](https://github.com/apache/pulsar/pull/15194)
- [Java] Fix performance regression with message listener [#15162](https://github.com/apache/pulsar/pull/15162)
- [Java] ConsumerBuilderImpl can not set null to deadLetterPolicy [#14980](https://github.com/apache/pulsar/pull/14980)
- [Java] Fix message publishing stuck when enabling batch [#14870](https://github.com/apache/pulsar/pull/14870)

> - [Java] Returns immutable data set when use TableView [#14833](https://github.com/apache/pulsar/pull/14833)

Reviewer 2: ```suggestion
- [Java] Returns immutable data set when using TableView [#14833](https://github.com/apache/pulsar/pull/14833)
```

- [C++] Fix single message metadata not set correctly [#15072](https://github.com/apache/pulsar/pull/15072)
- [C++] Add subscription properties to consumer for cpp [#15020](https://github.com/apache/pulsar/pull/15020)
- [C++] Fix connection is not closed when broker closes the connection to proxy[#14070](https://github.com/apache/pulsar/pull/14070)
- [C++] Fix send callback might not be invoked in key based batching [#14898](https://github.com/apache/pulsar/pull/14898)
- [C++] Fix C++ client compile error because of keyword optional is redundant in PaddingDemo.proto [#14862](https://github.com/apache/pulsar/pull/14862)
- [C++] Fix segmentation fault when creating socket failed [#14834](https://github.com/apache/pulsar/pull/14834)
- [C++] Fix the race condition of connect timeout task [#14823](https://github.com/apache/pulsar/pull/14823)
- [C++] Fix producer is never destructed until client is closed [#14797](https://github.com/apache/pulsar/pull/14797)
- [C++] Close messages_ when PartitionedConsumer is closed [#16887](https://github.com/apache/pulsar/pull/16887)
- [Python] Add getLastMessageIdAsync C binding [#16255](https://github.com/apache/pulsar/pull/16255)

> - [Python] Fixed reserved keys is not removed when JsonSchema being encoded [#15947](https://github.com/apache/pulsar/pull/15947)

Reviewer 2: ```suggestion
- [Python] Remove reserved keys when JsonSchema is being encoded [#15947](https://github.com/apache/pulsar/pull/15947)
```

- [Python] Use MacOS 10.15 as the target OS version for Python wheel files [#15788](https://github.com/apache/pulsar/pull/15788)

> - [Python] Removed dependencies only need by Python2 [#15460](https://github.com/apache/pulsar/pull/15460)

Reviewer 2: ```suggestion
- [Python] Remove dependencies only needed by Python 2 [#15460](https://github.com/apache/pulsar/pull/15460)
```

> - [Python] Removed scripts to create wheel files in vagrant VMs [#15459](https://github.com/apache/pulsar/pull/15459)

Reviewer 2: ```suggestion
- [Python] Remove scripts to create wheel files in vagrant VMs [#15459](https://github.com/apache/pulsar/pull/15459)
```

> - [Python] Adjusted script to build wheel for Python 3.7 on Mac [#15407](https://github.com/apache/pulsar/pull/15407)

Reviewer 2: ```suggestion
- [Python] Adjust scripts to build wheel files for Python 3.7 on Mac [#15407](https://github.com/apache/pulsar/pull/15407)
```

- [Python] Change python cp35-cp35m lib build version to manylinux1 [#15180](https://github.com/apache/pulsar/pull/15180)

> - [Python] Generate universal2 wheel files for MacOS [#15054](https://github.com/apache/pulsar/pull/15054)

Reviewer 2: ```suggestion
- [Python] Generate universal2 wheel files for macOS [#15054](https://github.com/apache/pulsar/pull/15054)
```

> - [Python] Added script to generate Wheel files for MacOS [#15024](https://github.com/apache/pulsar/pull/15024)

Reviewer 2: ```suggestion
- [Python] Add scripts to generate wheel files for macOS [#15024](https://github.com/apache/pulsar/pull/15024)
```

> - [Python] Added build for wheels on Alpine linux [#15016](https://github.com/apache/pulsar/pull/15016)

Reviewer 2: ```suggestion
- [Python] Add build for wheels on Alpine Linux [#15016](https://github.com/apache/pulsar/pull/15016)
```

> - [Python] Added build for Arm64 on Linux and Python 3.10 [#15004](https://github.com/apache/pulsar/pull/15004)

Reviewer 2: ```suggestion
- [Python] Add build for ARM64 on Linux and Python 3.10 [#15004](https://github.com/apache/pulsar/pull/15004)
```

### Pulsar IO and Pulsar Functions
- [Functions] Add the ability to customize logging level for Go & Python functions [#16939](https://github.com/apache/pulsar/pull/16939)
- [Functions] Use the schema set by the Function when it returns a Record [#17142](https://github.com/apache/pulsar/pull/17142)
- [Functions] Make mandatory to provide a schema in Context::newOutputRecordBuilder [#17118](https://github.com/apache/pulsar/pull/17118)
- [Functions] Fixed error when user starts with the pulsar functions local runner [#16565](https://github.com/apache/pulsar/pull/16565)
- [Functions] Fix netty.DnsResolverUtil compat issue on JDK9+ for the function Runtimes [#16423](https://github.com/apache/pulsar/pull/16423)

> - [Functions] Ensure bytes is a well-formed UTF-8 byte sequence when decode the FunctionState bytes to string [#16199](https://github.com/apache/pulsar/pull/16199)

Reviewer 2: ```suggestion
- [Functions] Ensure bytes are a well-formed UTF-8 byte sequence when decoding the FunctionState bytes to string [#16199](https://github.com/apache/pulsar/pull/16199)
```

- [Functions] Remove internal dependency: pulsar-functions-instance [#14925](https://github.com/apache/pulsar/pull/14925)
- [Functions] Allow a Function<GenericObject,?> to access the original Schema of the Message and use it [#14847](https://github.com/apache/pulsar/pull/14847)
- [Functions] Fix pulsar-managed runtimes failed start function with package URL from package management service [#14814](https://github.com/apache/pulsar/pull/14814)
- [Functions] Set function channel to idle to prevent DNS resolution of deleted pod [#14750](https://github.com/apache/pulsar/pull/14750)
- [Functions] Remove extra call to gi.stats.incrTotalProcessedSuccessfully() [#12316](https://github.com/apache/pulsar/pull/12316)
- [Functions] Fix python instance not process zip file correctly [#16697](https://github.com/apache/pulsar/pull/16697)
- [Functions] Missing assertion in KubernetesRuntimeTest::verifyJavaInstance [#16747](https://github.com/apache/pulsar/pull/16747)
- [IO Connector] IOConfigUtils support required and defaultValue annotations [#16785](https://github.com/apache/pulsar/pull/16785)

> - [IO Connector] Fixed kafka source config for pulsar-io-kafka when consumerConfigProperties="" [#16731](https://github.com/apache/pulsar/pull/16731)

Reviewer 2: ```suggestion
- [IO Connector] Fixed Kafka source config for pulsar-io-kafka when consumerConfigProperties is null [#16731](https://github.com/apache/pulsar/pull/16731)
```

- [Broker] Rename Pulsar schema metrics to specify OpenMetrics [#16610](https://github.com/apache/pulsar/pull/16610)
- [Broker] Rename Pulsar txn metrics to specify OpenMetrics [#16581](https://github.com/apache/pulsar/pull/16581)
- [Broker] Optimise msgOutCounter and bytesOutCounter [#16214](https://github.com/apache/pulsar/pull/16214)
- [Broker] Fail to expose managed ledger client stats to prometheus if bookkeeperClientExposeStatsToPrometheus is true [#16219](https://github.com/apache/pulsar/pull/16219)
- [Broker] Add metrics for pulsar web service thread pool [#14742](https://github.com/apache/pulsar/pull/14742)
- [Broker] Add support of PrometheusRawMetricsProvider for the Pulsar-Proxy [#14681](https://github.com/apache/pulsar/pull/14681)
- [Broker] Add filteredMsgCount for pulsar-admin broker-stats topics [#14531](https://github.com/apache/pulsar/pull/14531)
- [Broker] Improve /metrics endpoint performance [#14453](https://github.com/apache/pulsar/pull/14453)
- [Broker] Offloader metrics [#13833](https://github.com/apache/pulsar/pull/13833)
- [Broker] Improve /metrics endpoint performance [#14453](https://github.com/apache/pulsar/pull/14453)

> - [Broker] add metric for InMemoryDelayedDeliveryTracker's memory usage [#15867](https://github.com/apache/pulsar/pull/15867)

Reviewer 2: ```suggestion
- [Broker] Add metrics for InMemoryDelayedDeliveryTracker's memory usage [#15867](https://github.com/apache/pulsar/pull/15867)
```

- [Pulsar CLI] Add query options to the get topics in Namespace [#16167](https://github.com/apache/pulsar/pull/16167)
- [Pulsar CLI] Add cli cmd for subscription level dispatch-rate-limiter [#15862](https://github.com/apache/pulsar/pull/15862)
- [Pulsar CLI] Disable Pulsar client memory limit by default [#15748](https://github.com/apache/pulsar/pull/15748)
- [Pulsar CLI] Fix check on create function class name [#15700](https://github.com/apache/pulsar/pull/15862)
- [Pulsar CLI] Support filtering system topic when get list [#15410](https://github.com/apache/pulsar/pull/15410)
- [Pulsar CLI] Change the default for max-connections from 100 to 1 [#15387](https://github.com/apache/pulsar/pull/15387)
- [Pulsar CLI] Add short name for full name in admin cli [#14301](https://github.com/apache/pulsar/pull/14301)
- [Client Tool] Add tlsTrustCertFilePath as CLI argument for pulsar-client-tool [#16961](https://github.com/apache/pulsar/pull/16961)
- [Perf Tool] Able to set maxLookupRequest for pulsar-perf [#16967](https://github.com/apache/pulsar/pull/16967)

### Others
- [Proxy] Add TLS transport encryption for broker client [#16833](https://github.com/apache/pulsar/pull/16833)

> - [Proxy] Fix client service url [#16834](https://github.com/apache/pulsar/pull/16834)

Reviewer 2: ```suggestion
- [Proxy] Fix client service URL [#16834](https://github.com/apache/pulsar/pull/16834)
```

- [WebSocket] Support encryption in websocket proxy [#16234](https://github.com/apache/pulsar/pull/16234)
- [Package Management] Fix the new path /data introduced regression [#15367](https://github.com/apache/pulsar/pull/15367)
- [Schema] Add decode InputStream for Schema [#16659](https://github.com/apache/pulsar/pull/16659)
- [Schema] Expose timestamp field for SchemaData&SchemaInfo [#16380](https://github.com/apache/pulsar/pull/16380)
- [Schema] Add classLoader field for SchemaDefinition [#15915](https://github.com/apache/pulsar/pull/15915)
- [Schema] Fix conversion of TimestampMillisConversion has no effect when Jsr310Conversion enabled [#15863](https://github.com/apache/pulsar/pull/15863)
- [Schema] Ensure the deletion consistency of topic and schema [#14608](https://github.com/apache/pulsar/pull/14608)
- [Tiered Storage] Shared subscription: improvement with offloaded ledgers [#16417](https://github.com/apache/pulsar/pull/16417)
- [Tiered Storage] Fix the wrong secret key name get from env [#15814](https://github.com/apache/pulsar/pull/15814)
- [Tiered Storage] Add pure S3 provider for the offloader [#15710](https://github.com/apache/pulsar/pull/15710)

> - [Tiered Storage] Reduce cpu usage when offloading the ledger [#15063](https://github.com/apache/pulsar/pull/15063)

Reviewer 2: ```suggestion
- [Tiered Storage] Reduce CPU usage when offloading ledgers [#15063](https://github.com/apache/pulsar/pull/15063)
```

- [Tiered Storage] Fix potential NPE in MockManagedLedger [#15006](https://github.com/apache/pulsar/pull/15006)
- [Tiered Storage] Add API to scan objects on Tiered Storage [#14930](https://github.com/apache/pulsar/pull/14930)
- [Tiered Storage] Fix incorrect entryId in warn log when reading entry from tiered storage [#14685](https://github.com/apache/pulsar/pull/14685)
- Remove cmd-line level test retries [#16524](https://github.com/apache/pulsar/pull/16524)

> - Fix potential incompatible java opts caused by IS_JAVA_8 [#15444](https://github.com/apache/pulsar/pull/15444)

Reviewer 2: ```suggestion
- Fix potentially incompatible Java opts caused by IS_JAVA_8 [#15444](https://github.com/apache/pulsar/pull/15444)
```

- Remove obsolete jdk11 and errorprone-jdk11 maven profiles [#15505](https://github.com/apache/pulsar/pull/15505)
- Remove internal dependency: pulsar-functions-instance [#14925](https://github.com/apache/pulsar/pull/14925)
- Enable retries for apt-get when building Pulsar docker image [#14513](https://github.com/apache/pulsar/pull/15505)
- Migrate Docker images and CI to Java 17 [#14355](https://github.com/apache/pulsar/pull/14355)

### Library updates
- Upgrade to Jetty to 9.4.48.v20220622 [#16520](https://github.com/apache/pulsar/pull/16520)
- Bump jackson version from 2.13.2 to 2.13.3 [#16508](https://github.com/apache/pulsar/pull/16508)
- Bump maven-dependency-plugin to 3.3.0 [#16318](https://github.com/apache/pulsar/pull/16318)
- Bump os-maven-plugin version from 1.4.1.Final to 1.7.0 [#16308](https://github.com/apache/pulsar/pull/16308)
- Bump presto.version to run PrestoServer on JDK11+ [#16163](https://github.com/apache/pulsar/pull/16163)
- Update fastjson version to 1.2.83 [#16148](https://github.com/apache/pulsar/pull/16148)

> - Bump pyyaml from 5.3.1 to 5.4.1 to solve CVE-2020-14343 [#15989](https://github.com/apache/pulsar/pull/15989)

Reviewer 2: ```suggestion
- Bump PyYAML from 5.3.1 to 5.4.1 to solve CVE-2020-14343 [#15989](https://github.com/apache/pulsar/pull/15989)
```

- Bump spring version to 5.3.20 to solve CVE-2022-22970 [#15699](https://github.com/apache/pulsar/pull/15699)
- Bump snappy zstd version to fix CompressorCodecBackwardCompatTest on Apple M1 [#15698](https://github.com/apache/pulsar/pull/15698)
- Athenz: do not use uber-jar and bump to 1.10.50 to remove jackson-databind shaded dependency [#14884](https://github.com/apache/pulsar/pull/14884)
- Remove --illegal-access errors resulting from Google Guice - Pulsar IO, Offloaders and Pulsar SQL - Bump Guice to 5.1.0 [#14300](https://github.com/apache/pulsar/pull/14300)
- Bump prometheus client version from 0.5.0 to 0.15.0 [#13785](https://github.com/apache/pulsar/pull/13785)
- Upgrade log4j2 version to 2.18.0 [#16884](https://github.com/apache/pulsar/pull/16884

### Documentation

> This section only highlights the non-2.11.0-specific doc availability that fulfills existing content gaps.

Reviewer 2: ```suggestion
This section only highlights the availability of docs serving existing features.
```

- Add an introductory table about [BookKeeper recovery metrics](https://pulsar.apache.org/docs/next/reference-metrics#replication-metrics) [#16554](https://github.com/apache/pulsar/pull/16554)
- Publish REST API docs for [lookup](https://pulsar.apache.org/lookup-rest-api/?version=master) related operations [#16621](https://github.com/apache/pulsar/pull/16621)
- Add more concepts and user tasks about [bookie isolation](https://pulsar.apache.org/docs/next/administration-isolation-bookie) [#16843](https://github.com/apache/pulsar/pull/16843)
- Use 2.8.x/2.9.x/2.10.x doc set instead of version-specific doc set [#17074](https://github.com/apache/pulsar/pull/17074)

> For the comprehensive list of doc-related improvements in 2.11.0, you can go to [GitHub](https://github.com/apache/pulsar/pulls?q=is%3Apr+milestone%3A2.11.0+label%3Adoc+is%3Aclosed). A special thanks to the following contributors (alphabetic Github IDs) who helped add docs for Pulsar 2.11.0.

> > [704237006](https://github.com/704237006), [AlphaWang](https://github.com/AlphaWang), [Anonymitaet](https://github.com/Anonymitaet), [asafm](https://github.com/asafm), [cbornet](https://github.com/cbornet), [codelipenghui](https://github.com/codelipenghui), [coderzc](https://github.com/coderzc), [Demogorgon314](https://github.com/Demogorgon314), [ericsyh](https://github.com/ericsyh), [fantapsody](https://github.com/fantapsody), [futeng](https://github.com/futeng), [hangc0276](https://github.com/hangc0276), [heesung-sn](https://github.com/heesung-sn), [hrsakai](https://github.com/hrsakai), [HQebupt](https://github.com/HQebupt), [horizonzy](https://github.com/horizonzy), [Huanli-Meng](https://github.com/Huanli-Meng), [gaozhangmin](https://github.com/gaozhangmin), [ikilobyte](https://github.com/ikilobyte), [Jason918](https://github.com/Jason918), [komalatammal](https://github.com/komalatammal), [larshp](https://github.com/larshp), [lgxbslgx](https://github.com/lgxbslgx), [lhotari](https://github.com/lhotari), [liangyepianzhou](https://github.com/liangyepianzhou), [liudezhi2098](https://github.com/liudezhi2098), [liuzhuang2017](https://github.com/liuzhuang2017), [Mans2singh](https://github.com/mans2singh), [massakam](https://github.com/massakam), [mattisonchao](https://github.com/mattisonchao), [maxsxu](https://github.com/maxsxu), [mendonk](mendonk), [merlimat](https://github.com/merlimat), [michaeljmarshall](https://github.com/michaeljmarshall), [misselvexu](https://github.com/misselvexu), [momo-jun](https://github.com/momo-jun), [MTwz](https://github.com/MTwz), [nahguam](https://github.com/nahguam),[nicoloboschi](https://github.com/nicoloboschi), [nodece](https://github.com/nodece), [poorbarcode](https://github.com/poorbarcode), [rdhabalia](https://github.com/rdhabalia), [RobertIndie](https://github.com/RobertIndie), [samredai](https://github.com/samredai), [Shawyeok](https://github.com/Shawyeok), [Sherlock113](https://github.com/Sherlock113), [shibd](https://github.com/shibd), [Shoothzj](https://github.com/Shoothzj), [SignorMercurio](https://github.com/SignorMercurio), [Technoboy-](https://github.com/Technoboy-), [tjiuming](https://github.com/tjiuming), [tisonkun](https://github.com/tisonkun), [urfreespace](https://github.com/urfreespace), [wangjialing218](https://github.com/wangjialing218), [wenbingshen](https://github.com/wenbingshen), [youzipi](https://github.com/youzipi), [zwOvO](https://github.com/zwOvO).

Reviewer 2: ```suggestion
> [704237006](https://github.com/704237006), [AlphaWang](https://github.com/AlphaWang), [Anonymitaet](https://github.com/Anonymitaet), [asafm](https://github.com/asafm), [cbornet](https://github.com/cbornet), [codelipenghui](https://github.com/codelipenghui), [coderzc](https://github.com/coderzc), [Demogorgon314](https://github.com/Demogorgon314), [ericsyh](https://github.com/ericsyh), [fantapsody](https://github.com/fantapsody), [futeng](https://github.com/futeng), [hangc0276](https://github.com/hangc0276), [heesung-sn](https://github.com/heesung-sn), [hrsakai](https://github.com/hrsakai), [HQebupt](https://github.com/HQebupt), [horizonzy](https://github.com/horizonzy), [Huanli-Meng](https://github.com/Huanli-Meng), [gaozhangmin](https://github.com/gaozhangmin), [ikilobyte](https://github.com/ikilobyte), [Jason918](https://github.com/Jason918), [komalatammal](https://github.com/komalatammal), [labuladong](https://github.com/labuladong), [larshp](https://github.com/larshp), [lgxbslgx](https://github.com/lgxbslgx), [lhotari](https://github.com/lhotari), [liangyepianzhou](https://github.com/liangyepianzhou), [liudezhi2098](https://github.com/liudezhi2098), [liuzhuang2017](https://github.com/liuzhuang2017), [Mans2singh](https://github.com/mans2singh), [massakam](https://github.com/massakam), [mattisonchao](https://github.com/mattisonchao), [maxsxu](https://github.com/maxsxu), [mendonk](mendonk), [merlimat](https://github.com/merlimat), [michaeljmarshall](https://github.com/michaeljmarshall), [misselvexu](https://github.com/misselvexu), [momo-jun](https://github.com/momo-jun), [MTwz](https://github.com/MTwz), [nahguam](https://github.com/nahguam),[nicoloboschi](https://github.com/nicoloboschi), [nodece](https://github.com/nodece), [poorbarcode](https://github.com/poorbarcode), [Pradhyumnakashyap](https://github.com/pradhyumnakashyap), [rdhabalia](https://github.com/rdhabalia), [RobertIndie](https://github.com/RobertIndie), [samredai](https://github.com/samredai), [Shawyeok](https://github.com/Shawyeok), [Sherlock113](https://github.com/Sherlock113), [shibd](https://github.com/shibd), [Shoothzj](https://github.com/Shoothzj), [SignorMercurio](https://github.com/SignorMercurio), [Technoboy-](https://github.com/Technoboy-), [tjiuming](https://github.com/tjiuming), [tisonkun](https://github.com/tisonkun), [urfreespace](https://github.com/urfreespace), [wangjialing218](https://github.com/wangjialing218), [wenbingshen](https://github.com/wenbingshen), [youzipi](https://github.com/youzipi), [zwOvO](https://github.com/zwOvO).
```

Reviewer 2: @Technoboy- I added two more doc contributors.
